### DISCUSSION TITLE
Versioning of named linter groups ('builtinLinters' option)

### DISCUSSION THREAD

**Discussion by fredemmott** [hhvm/hhast#425] [General]
Should a user be able to say "I started using HHAST at vx.y, don't give me any new errors even when I upgrade?"

--

Should users be able to provide separate version constraints for different categories of linters?

--

What if a bug is fixed in a linter and it now finds more cases of the same problem?

**Comment by fredemmott** [comment]
> linter groups

Perhaps we should rethink our core abstractions a bit more here - should it be possible for a linter to *be* a linter group?

For example, should `hhclientlinter` be a linter group, and a specific way to run them?

Perhaps linters have a 'lintimplementation' - something like hhclientlinter is both a lintergroup and a lintimplementation, with several codes. most linters are lintergroups with single codes, and a single dedicated lintimplementation?

I don't know if unifying these concepts actually makes sense, but thought it was worth raising - I think @atry is most likely to have a good feel for this idea is good.

**Comment by Atry** [comment]
I think we should [separate policy from mechanism; separate interfaces from engines](http://www.catb.org/~esr/writings/taoup/html/ch01s06.html#id2877777). Essentially, the linter engine requires the following function:

``` hack
type ErrorCode = int | classname<SingleRuleLinter>;
function shouldReportErrorFromFile(ErrorCode $error_code, string $path): bool;
```

And we can provide some policy functions to group error codes:

``` hack
function isCodeSmell(ErrorCode $error_code): bool;
function isPHPFlavored(ErrorCode $error_code): bool;
function isFacebookFlavored(ErrorCode $error_code): bool;
```

If we could allow users to create the `isErrorCodeForFile` in Hack, it should be some simple combinations of bool operators and regex matching against the path.

**Reply by fredemmott** [reply]
I think we need a separate discussion for what the groups are and how to implement them - this isn't versinoing (which is also why I marked my previous comment off-topic) :)

**Reply by fredemmott** [reply]
ah, https://github.com/hhvm/hhast/discussions/428

**Comment by fredemmott** [comment]
Tying into lexidor's comment at https://github.com/hhvm/hhast/discussions/423 : 

Should we have minimum and *maximum* versions?

For example, if I want `fb-style: 4.133`, does that mean I *require* 4.133, or I *permit* any linter up to and including the 4.133 set?

Do I need to be able to explicitly say "fb-style 4.130 or above is required, but anything up to 4.133 is ok"?

--

I think the lower bound is implied by the overall hhast version constraint and CI environment, and we don't need a separate setting/constraint: if you permit hhast 4.130 and fb-style 4.133, that implies that you're OK with 4.130 doing the best it can.

i.e., the version constraint here is a maximum, not a minimum

**Reply by lexidor** [reply]
I believe there is a good reason to allow for a minimum version. If a buggy linter is introduced and later fixed, you want to exclude the buggy version of this linter. This is  it a constraint in the composer sense (exits with a failed exit code if unmet), but a gentle way to tell hhast to skip this buggy linter.

**Reply by fredemmott** [reply]
If we distinguish at a configuration level between linters and groups, is that needed for groups, or just specific linters?

**Reply by lexidor** [reply]
`fb-style: "4.133", DontAwaitInALoopLinter: "<4.90"`.

**Reply by lexidor** [reply]
Minimum versions on groups are indeed far less useful. Having this be a per linter only thing wouldn't be that bad.

**Comment by fredemmott** [comment]
> Should a user be able to say "I started using HHAST at vx.y, don't give me any new errors even when I upgrade?"

> What if a bug is fixed in a linter and it now finds more cases of the same problem?

I think we should consider this acceptable, out of practicality or nothing else: otherwise we need to maintain bug-for-bug compatibility and versioning within a linter.

In other words, we need to interpret version constraints as saying "I'm interested in this class of errors", rather than "I only want the same exact same error instances reported that HHAST x.y did"
